internal/xds/translator/ext_authz.go

```diff
@@ -0,0 +1,291 @@
+	return fmt.Sprintf("%s_%s", extAuthzFilter, route.Name)
+}
+
+func extAuthzConfig(route *ir.HTTPRoute) (*extauthzv3.ExtAuthz, error) {
+	grpcURI := route.ExtAuthz.GRPCURI
+	grpc, err := url2Cluster(grpcURI, false)
+	if err != nil {
+		return nil, err
+	}
+
+	authProto := &extauthzv3.ExtAuthz{
+		Services: &extauthzv3.ExtAuthz_GrpcService{
+			GrpcService: &corev3.GrpcService{
+				TargetSpecifier: &corev3.GrpcService_EnvoyGrpc_{
+					EnvoyGrpc: &corev3.GrpcService_EnvoyGrpc{ClusterName: grpc.name},
+				},
+			},
+		},
+		TransportApiVersion: corev3.ApiVersion_V3,
+		FailureModeAllow:    false, // do not let the request pass when authz unavailable
+		StatusOnError:       &typev3.HttpStatus{Code: typev3.StatusCode_ServiceUnavailable},
+	}
+	return authProto, nil
+}
+
+// routeContainsExtAuthz returns true if external authorizations exists for the
+// provided route.
+func routeContainsExtAuthz(irRoute *ir.HTTPRoute) bool {
+	if irRoute == nil {
+		return false
+	}
+
+	if irRoute != nil &&
+		irRoute.ExtAuthz != nil &&
+		irRoute.ExtAuthz.GRPCURI != "" {
+		return true
+	}
+
+	return false
+}
+
+// patchResources creates external authorizations clusters from the provided routes, if needed.
+func (*extAuth) patchResources(tCtx *types.ResourceVersionTable, routes []*ir.HTTPRoute) error {
+	if tCtx == nil || tCtx.XdsResources == nil {
+		return errors.New("xds resource table is nil")
+	}
+
+	var errs error
+	for _, route := range routes {
+		if !routeContainsExtAuthz(route) {
+			continue
+		}
+
+		var (
+			grpc    *urlCluster
+			ds      *ir.DestinationSetting
+			tSocket *corev3.TransportSocket
+			err     error
+		)
+
+		url := route.ExtAuthz.GRPCURI
+		grpc, err = url2Cluster(url, false)
+		if err != nil {
+			errs = multierror.Append(errs, err)
+			continue
+		}
+
+		ds = &ir.DestinationSetting{
+			Weight:    ptr.To(uint32(1)),
+			Endpoints: []*ir.DestinationEndpoint{ir.NewDestEndpoint(grpc.hostname, grpc.port)},
+			Protocol:  ir.GRPC,
+		}
+
+		if grpc.tls {
+			// grpcURI is using TLS gRPC HT
+			tSocket, err = buildExtAuthzTLSocket()
+			if err != nil {
+				errs = multierror.Append(errs, err)
+				continue
+			}
+		}
+
+		if err = addXdsCluster(tCtx, &xdsClusterArgs{
+			name:         grpc.name,
+			settings:     []*ir.DestinationSetting{ds},
+			tSocket:      tSocket,
+			endpointType: grpc.endpointType,
+		}); err != nil && !errors.Is(err, ErrXdsClusterExists) {
+			errs = multierror.Append(errs, err)
+		}
+	}
+
+	return errs
+}
+
+// buildExtAuthzTLSocket builds the TLS socket for the ext authx.
+func buildExtAuthzTLSocket() (*corev3.TransportSocket, error) {
+	tlsCtxProto := &tlsv3.UpstreamTlsContext{
+		CommonTlsContext: &tlsv3.CommonTlsContext{
+			ValidationContextType: &tlsv3.CommonTlsContext_ValidationContext{
+				ValidationContext: &tlsv3.CertificateValidationContext{
+					TrustedCa: &corev3.DataSource{
+						Specifier: &corev3.DataSource_Filename{
+							Filename: envoyTrustBundle,
```

Comment:
Do we need to support self-signed CAs?

Reply:
eventually yes, we'll need a `tls` setting